File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # prev, curr = None, head

        # while curr:
        #     temp = curr.next
        #     curr.next = prev
        #     prev = curr
        #     curr = temp
            

        # return prev

        if not head: return None
        
        newHead = head

        if head.next:
            newHead = self.reverseList(newHead.next)
            head.next.next = head
        head.next = None

        return newHead

File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while(curr!=null){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;

        // if(!head) return null;

        // ListNode* newHead = head;

        // if(head.next){
        //     newHead = reverseList(head.next);
        //     head.next.next = head;
        // }

        // head.next = nullptr;

        // return newHead;

    }
}